### PR TITLE
Salt's ansiblegate module fixes

### LINKED ISSUE #59792: [BUG] 3000.8 3002.5 RC13003 ansiblegate if it is able to run doesn't pass args or kwargs to the submodule.
**Description**
If you are able to get ansible gate to run. which is a big if right now. It doesn't seem to work with 3002 or 3003 as it is not passing the value of any args to the system.module. and has str errors in 3000 

**Setup**
```
salt-call ansible.system.ping 
```

will return ping: pong under normal circumstances. 

but if you pass data='string' it will respond with string. 

however currently if you pass data='hello' to ansible.system.ping you will still get pong. 

After putting some logging into the ansiblegate module that showed what the args values were. they were blank dicts. nothing in them. 

**Expected behavior**
args passed to ansiblegate are passed to the ansible modules

### LINKED ISSUE #60207: [BUG] Ansiblegate is not usable after Salt 3001 release
**Description**
Prior to the 3001 release of salt, the ansiblegate module worked by adding the ansible functions into the loaded salt module as a wrapper.

```
[root@b9dbd4ce1076 testing]# salt-call -l debug --local ansible.system.ping
[DEBUG   ] Missing configuration file: /etc/salt/minion
[DEBUG   ] Using cached minion ID from /etc/salt/minion_id: b9dbd4ce1076
[DEBUG   ] Configuration file path: /etc/salt/minion
[WARNING ] Insecure logging configuration detected! Sensitive data may be logged.
[DEBUG   ] Grains refresh requested. Refreshing grains.
[DEBUG   ] Missing configuration file: /etc/salt/minion
[DEBUG   ] Could not determine init system from command line: (bash)
[DEBUG   ] The `lspci` binary is not available on the system. GPU grains will not be available.
[DEBUG   ] LazyLoaded zfs.is_supported
[DEBUG   ] Determining pillar cache
[DEBUG   ] LazyLoaded jinja.render
[DEBUG   ] LazyLoaded yaml.render
[DEBUG   ] LazyLoaded jinja.render
[DEBUG   ] LazyLoaded yaml.render
[DEBUG   ] Resolving Ansible modules
[DEBUG   ] Installing Ansible modules
[DEBUG   ] LazyLoaded ansible.system.ping
[DEBUG   ] LazyLoaded direct_call.execute

Passed invalid arguments: a bytes-like object is required, not 'str'.

Usage:
Available
Traceback (most recent call last):
  File "/usr/local/lib/python3.6/site-packages/salt/cli/caller.py", line 218, in call
    ret['return'] = self.minion.executors[fname](self.opts, data, func, args, kwargs)
  File "/usr/local/lib/python3.6/site-packages/salt/executors/direct_call.py", line 12, in execute
    return func(*args, **kwargs)
  File "/usr/local/lib/python3.6/site-packages/salt/modules/ansiblegate.py", line 210, in _cmd
    return _caller.call(cmd_name, *args, **kwargs)
  File "/usr/local/lib/python3.6/site-packages/salt/modules/ansiblegate.py", line 167, in call
    stdin=proc_out.stdout, stdout=subprocess.PIPE, timeout=self.timeout)
  File "/usr/local/lib/python3.6/site-packages/salt/utils/timed_subprocess.py", line 36, in __init__
    self.stdin = salt.utils.stringutils.to_bytes(self.stdin.replace('\\n', '\n'))
TypeError: a bytes-like object is required, not 'str'
```
Although there's a traceback, the ansible modules were properly injected into Salt's `ansible` module.

Now, for Salt 3001+
```
[root@b9dbd4ce1076 testing]# salt-call -l debug --local ansible.system.ping
[DEBUG   ] Missing configuration file: /etc/salt/minion
[DEBUG   ] Using cached minion ID from /etc/salt/minion_id: b9dbd4ce1076
[DEBUG   ] Configuration file path: /etc/salt/minion
[WARNING ] Insecure logging configuration detected! Sensitive data may be logged.
[DEBUG   ] Grains refresh requested. Refreshing grains.
[DEBUG   ] Missing configuration file: /etc/salt/minion
[DEBUG   ] Elapsed time getting FQDNs: 0.11789274215698242 seconds
[DEBUG   ] Could not determine init system from command line: (bash)
[DEBUG   ] The `lspci` binary is not available on the system. GPU grains will not be available.
[DEBUG   ] LazyLoaded zfs.is_supported
[DEBUG   ] Determining pillar cache
[DEBUG   ] LazyLoaded jinja.render
[DEBUG   ] LazyLoaded yaml.render
[DEBUG   ] LazyLoaded jinja.render
[DEBUG   ] LazyLoaded yaml.render
[DEBUG   ] Resolving Ansible modules
[DEBUG   ] Installing Ansible modules
[DEBUG   ] Missing configuration file: /etc/salt/minion
[DEBUG   ] Using cached minion ID from /etc/salt/minion_id: b9dbd4ce1076
[DEBUG   ] Could not LazyLoad boto3.assign_funcs: 'boto3.assign_funcs' is not available.
[DEBUG   ] Error loading module.boto3_elasticsearch: __init__ failed
Traceback (most recent call last):
  File "/usr/local/lib/python3.6/site-packages/salt/loader.py", line 1754, in _load_module
    module_init(self.opts)
  File "/usr/local/lib/python3.6/site-packages/salt/modules/boto3_elasticsearch.py", line 96, in __init__
    __utils__["boto3.assign_funcs"](__name__, "es")
  File "/usr/local/lib/python3.6/site-packages/salt/loader.py", line 1278, in __getitem__
    func = super(LazyLoader, self).__getitem__(item)
  File "/usr/local/lib/python3.6/site-packages/salt/utils/lazy.py", line 108, in __getitem__
    raise KeyError(key)
KeyError: 'boto3.assign_funcs'
[DEBUG   ] DSC: Only available on Windows systems
[DEBUG   ] Module PSGet: Only available on Windows systems
[DEBUG   ] Could not LazyLoad ansible.system.ping: 'ansible.system.ping' is not available.
'ansible.system.ping' is not available.
```

### PR DESCRIPTION
**Pull request by s0undt3ch**
### What does this PR do?
* Inject the ansible functions into Salt's ansiblegate module.
This was broken in 3001(https://github.com/saltstack/salt/commit/92438882a94cbd4b1f1ec94eeda0bf1a35b98d27#diff-519146c38c0ed7bc961ffc5f408d3dd380abb8760d48e839743dbd38dc28ba15R323).
* Properly pass keyword arguments to ansible modules

### What issues does this PR fix or reference?
Fixes #60207
Fixes #59792

### Merge requirements satisfied?
**[NOTICE] Bug fixes or features added to Salt require tests.**

- [x] Docs
- [x] Changelog - https://docs.saltstack.com/en/master/topics/development/changelog.html
- [x] Tests written/updated

### Commits signed with GPG?
Yes

### CODE REVIEW

**Review by krionbsd — DISMISSED** [review]

**Review by Ch3LL — DISMISSED** [review]

File: salt/modules/ansiblegate.py [hunk 1]
```diff
@@ -222,99 +58,202 @@ def _set_function(cmd_name, doc):
...
+    else:
+        _kwargs = {k: v for (k, v) in kwargs.items() if not k.startswith("__pub")}
+
+    for key, value in _kwargs.items():
+        module_args.append("{}={}".format(key, salt.utils.json.dumps(value)))
+
+    with NamedTemporaryFile(mode="w") as inventory:
+
+        ansible_binary_path = salt.utils.path.which("ansible")
+        log.debug("Calling ansible module %r with arguments %r", module, module_args)
```
Comment (Ch3LL) [inline salt/modules/ansiblegate.py:207]: Would it be possible for this to log any sensitive information, like passwords?

Comment (s0undt3ch) [reply salt/modules/ansiblegate.py:207]: If ansible allows passing passwords as arguments, yes.
Don't really know.

Take that log out?

Comment (Ch3LL) [reply salt/modules/ansiblegate.py:207]: Yeah, or include: `log.debug("Calling ansible module %r ", module)` If that is still of use.

Comment (s0undt3ch) [reply salt/modules/ansiblegate.py:207]: :+1:

**Review by Ch3LL — DISMISSED** [review]

File: tests/pytests/integration/states/test_ansiblegate.py [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+                        "-o StrictHostKeyChecking=false "
+                        "-o UserKnownHostsFile=/dev/null "
+                    ),
+                },
+            },
+        },
+    }
+    with salt.utils.files.fopen(str(inventory), "w") as yaml_file:
+        yaml.dump(data, yaml_file, default_flow_style=False)
+    return str(inventory)
```
Comment (waynew) [inline tests/pytests/integration/states/test_ansiblegate.py:58]: **suggestion** we `str(inventory)` twice, prefer doing it earlier.

File: tests/pytests/unit/modules/test_ansiblegate.py [hunk 3]
```diff
@@ -1,200 +1,134 @@
...
-    """
-    with patch("salt.modules.ansiblegate.ansible", None):
-        assert ansible.__virtual__() == "ansible"
+    with subtests.test("missing ansible-playbook binary"):
+        with patch(
+            "salt.utils.path.which",
+            side_effect=[
+                "/path/to/ansible",
+                "/path/to/ansible-doc",
+                "/path/to/ansible-playbook",
```
Comment (waynew) [inline tests/pytests/unit/modules/test_ansiblegate.py:75]: **question**: is this a misnamed subtest? Says missing playbook, but it's right here in the patch 🤔

File: tests/pytests/unit/states/test_ansiblegate.py [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+    """
+
+    success_output = json.loads(
+        playbooks_examples_dir.joinpath("success_example.json").read_text()
+    )
+
+    with patch.dict(
+        ansiblegate.__salt__,
+        {"ansible.playbooks": MagicMock(return_value=success_output)},
+    ), patch("salt.utils.path.which", MagicMock(return_value=True)):
```
Comment (waynew) [inline tests/pytests/unit/states/test_ansiblegate.py:32]: **suggestion** just `patch(..., return_value=True)`

File: tests/pytests/unit/states/test_ansiblegate.py [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+    Test ansible.playbooks failed states executions.
+    :return:
+    """
+    failed_output = json.loads(
+        playbooks_examples_dir.joinpath("failed_example.json").read_text()
+    )
+    with patch.dict(
+        ansiblegate.__salt__,
+        {"ansible.playbooks": MagicMock(return_value=failed_output)},
+    ), patch("salt.utils.path.which", MagicMock(return_value=True)):
```
Comment (waynew) [inline tests/pytests/unit/states/test_ansiblegate.py:55]: **suggestion** also applies here

**Review by waynew — CHANGES_REQUESTED** [review]
Just a couple of minor tweaks, but... Without actually running locally, this looks great!

File: salt/modules/ansiblegate.py [hunk 6]
```diff
@@ -222,99 +58,202 @@ def _set_function(cmd_name, doc):
...
+    if proc.returncode != 0:
+        return (
+            False,
+            "Failed to get the listing of ansible modules:\n{}".format(proc.stderr),
+        )
+
+    ansible_module_listing = salt.utils.json.loads(proc.stdout)
+    for key in list(ansible_module_listing):
+        if key.startswith("ansible."):
+            alias = key.replace("ansible.", "")
```
[the remark this answers is not in the document]
Comment (s0undt3ch) [reply salt/modules/ansiblegate.py:105]: Are you also suggesting taking out `if key.startswith("ansible."):` ?

File: salt/modules/ansiblegate.py [hunk 7]
```diff
@@ -222,99 +58,202 @@ def _set_function(cmd_name, doc):
...
+
+    for key, value in _kwargs.items():
+        module_args.append("{}={}".format(key, salt.utils.json.dumps(value)))
+
+    with NamedTemporaryFile(mode="w") as inventory:
+
+        ansible_binary_path = salt.utils.path.which("ansible")
+        log.debug("Calling ansible module %r", module)
+        try:
+            proc_exc = subprocess.run(
```
[the remark this answers is not in the document]
Comment (s0undt3ch) [reply salt/modules/ansiblegate.py:209]: > why subprocess.run directly?

Don't really want or need the whole `cmdmod` machinery for this.

> Does --json not work here?

If I recall correctly, no, plus, the output is already in JSON, given that is has additional output we need to ignore in order to be able to load said JSON.

**Review by Ch3LL — APPROVED** [review]

**Review by krionbsd — APPROVED** [review]

**Review by garethgreenaway — APPROVED** [review]

**Pull request merged**